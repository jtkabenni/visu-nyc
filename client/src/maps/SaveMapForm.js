import React, { useState, useContext, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import UserContext from "../auth/UserContext";
/** Form for creating a new snack or drink item.
 *
 * Has state for the name/quantity of the item; on submission,
 * sends {name, qty} to fn rec'd from parent.
 *
 */

const SaveMapForm = ({ saveMap }) => {
  const { user } = useContext(UserContext);

  const INITIAL_STATE = {
    name: "",
    file: null,
    notes: "",
  };
  const navigate = useNavigate();
  const [formData, setFormData] = useState(INITIAL_STATE);
  const [formErrors, setFormErrors] = useState([]);
  /** Send {name, quantity} to parent
   *    & clear form. */

  async function handleSubmit(evt) {
    evt.preventDefault();
    let result = await saveMap(formData);
    if (result.success) {
      navigate("/profile");
    } else {
      setFormErrors(result.errors);
    }
  }

  /** Update local state w/curr state of input elem */

  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData((fData) => ({
      ...fData,
      [name]: value,
    }));
  };

  const handleUpload = (evt) => {
    evt.persist();

    setFormData((fData) => ({
      ...fData,
      file: evt.target.files[0],
    }));
  };

  useEffect(() => {
    if (!user) {
      navigate("/");
    }
  }, [user, navigate]);
  /** render form */
  return (
    <div class="form-container">
      <form onSubmit={handleSubmit}>
        <h3>Add a new map</h3>

        <input
          placeholder="Name"
          id="name"
          name="name"
          value={formData.name}
          onChange={handleChange}
        />

        <input
          placeholder="File"
          type="file"
          id="file"
          name="file"
          onChange={handleUpload}
        />

        <input
          placeholder="Notes"
          id="notes"
          name="notes"
          value={formData.notes}
          onChange={handleChange}
        />
        {formErrors.map((e) => (
          <p className="error">{e}</p>
        ))}
        <button>Save</button>
      </form>
    </div>
  );
};

export default SaveMapForm;
