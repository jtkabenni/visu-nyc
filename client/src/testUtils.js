import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  first_name: "testfirst",
  last_name: "testlast",
  email: "test@test.net",
  is_admin: true,
  auth_type: "custom",
  maps: [],
};

const demoCurrUserMaps = [];

const UserProvider = ({
  children,
  currUser = demoUser,
  currUserMaps = demoCurrUserMaps,
}) => (
  <UserContext.Provider value={{ currUser, currUserMaps }}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };
