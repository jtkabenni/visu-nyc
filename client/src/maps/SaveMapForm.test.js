import React from "react";
import { render } from "@testing-library/react";
import SaveMapForm from "./SaveMapForm";
import { UserProvider } from "../testUtils";
import { BrowserRouter as Router } from "react-router-dom";

it("matches snapshot", function () {
  const { asFragment } = render(
    <Router>
      <UserProvider>
        <SaveMapForm />
      </UserProvider>
    </Router>
  );
  expect(asFragment()).toMatchSnapshot();
});
