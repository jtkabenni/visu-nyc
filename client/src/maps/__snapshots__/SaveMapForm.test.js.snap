// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="form-container"
  >
    <form>
      <h3>
        Add a new map
      </h3>
      <input
        id="name"
        name="name"
        placeholder="Name"
        value=""
      />
      <input
        id="file"
        name="file"
        placeholder="File"
        type="file"
      />
      <input
        id="notes"
        name="notes"
        placeholder="Notes"
        value=""
      />
      <button>
        Save
      </button>
    </form>
  </div>
</DocumentFragment>
`;
